File: nodejs/5/segment.mjs
import {toInt} from "../lib/numbers.mjs";
import {range, zip} from "../lib/lists.mjs";

export class Segment {
  x1;
  y1;
  x2;
  y2;

  constructor(x1, y1, x2, y2) {
    this.x1 = x1;
    this.y1 = y1;
    this.x2 = x2;
    this.y2 = y2;
  }

  static from(line) {
    const [_, x1, y1, x2, y2, ...__] = line.match(/^(\d+),(\d+) -> (\d+),(\d+)$/);
    return new Segment(toInt(x1), toInt(y1), toInt(x2), toInt(y2));
  }

  isDiagonal() {
    return !this.isVertical() && !this.isHorizontal();
  }

  isVertical() {
    return this.x1 === this.x2;
  }

  isHorizontal() {
    return this.y1 === this.y2;
  }

  expandCoords() {
    if (this.isHorizontal())
      return range(this.x1, this.x2).map(x => [x, this.y1]);

    if (this.isVertical())
      return range(this.y1, this.y2).map(y => [this.x1, y]);

    return zip(range(this.x1, this.x2), range(this.y1, this.y2))
  }
}

File: java/app/src/main/java/com/github/ggalmazor/aoc2021/day5/Segment.java
package com.github.ggalmazor.aoc2021.day5;

import static com.github.ggalmazor.aoc2021.lib.Lists.range;
import static com.github.ggalmazor.aoc2021.lib.Lists.zip;

import java.util.List;

public record Segment(Coordinates a, Coordinates b) {
  public static Segment parse(String line) {
    String[] split = line.split(" -> ");
    Coordinates a = Coordinates.parse(split[0]);
    Coordinates b = Coordinates.parse(split[1]);
    return new Segment(a, b);
  }

  public List<Coordinates> expand() {
    if (isHorizontal())
      return range(a.x(), b.x(), true).stream().map(x -> new Coordinates(x, a.y())).toList();

    if (isVertical())
      return range(a.y(), b.y(), true).stream().map(y -> new Coordinates(a.x(), y)).toList();

    return zip(range(a.x(), b.x(), true), range(a.y(), b.y(), true)).stream().map(pair -> new Coordinates(pair.left(), pair.right())).toList();
  }

  private boolean isHorizontal() {
    return a.y() == b.y();
  }

  private boolean isVertical() {
    return a.x() == b.x();
  }

  public boolean isDiagonal() {
    return !isHorizontal() && !isVertical();
  }
}
